intellij/src/saros/intellij/filesystem/IntelliJProjectImpl.java

```diff
@@ -523,9 +527,18 @@ public VirtualFile findVirtualFile(final IPath path) {
     return null;
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * <p>This method operates under the assumption that disposed {@link Project} objects still allow
+   * requests for the project name. This method needs to be adjusted once this assumption no longer
+   * holds.
+   *
+   * @return a hash code value for this object
+   */
   @Override
   public int hashCode() {
-    return moduleName.hashCode();
+    return Objects.hash(project.getName(), moduleName);
```

Comment:
Looking at https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/openapi/project/impl/ProjectImpl.java#L126 it is possible to change the name. Of course this is a corner case (and I have not checked if it is possible in Eclipse too, but I guess all objects are immutable) and I do not even know when this method is called (renaming the project ?).

The default ProjectImpl class also have no equals and hashcode override, so the question is: is there only one project object available for a given project (if yes just use hash on the project object), otherwise we have to find another method that always returns the same value. However the only method that is NotNullable is getLocationHash which itself uses getName ...

Reply:
*Currently* project objects *seem* to be singleton from my limited testing. But I could not find any documentation explicitly mentioning this (see [project documentation](https://www.jetbrains.org/intellij/sdk/docs/basics/project_structure.html)).

I am not sure whether basing our logic on this assumption is the best idea, especially since this assumption is not easy to check. E.g. if the logic were to be changed tomorrow so that the project objects were just cached for a while but periodically re-generated, I am not sure how we could notice this internal change without a lot of manual testing/upstream source code searching. It is hard to construct comprehensive unit-tests for this. Furthermore, we are not always getting the project from the same source (sometimes we get it from the module, other times it is given to us by a listener, etc.), which makes testing all cases even harder.

Or am I just overthinking the whole thing?

Another option would be to handle this the same way we are handling moved resources, i.e. adding a listener that removes and re-adds the projec/module with the new hash code when it was moved. But we would have to apply this recursively as the module hashcode is based on the project hashcode and the file/folder hashcode is based on the module hashcode.

Reply:
>Of course this is a corner case (and I have not checked if it is possible in Eclipse too, but I guess all objects are immutable) and I do not even know when this method is called (renaming the project ?).

You can just rename the project in the [settings](https://stackoverflow.com/a/21180363/11010012), which will immediately be reflected by the project object. Sadly, I could not find a listener for this change.

The change creates the file `.idea/.name` if the name differs from the name of the project base directory, so we could theoretically react to the creation/modification of the file and react with updating the hashset, but I am not sure whether this is the best solution.

Reply:
I could not find think of a better solution approach...

I posted a question regarding the topic to the IntelliJ forum: https://intellij-support.jetbrains.com/hc/en-us/community/posts/360004300240-Immutable-hashcode-for-project-and-module-objects

Hopefully somebody there (e.g. somebody of the IntelliJ development team) can suggest a better approach or confirm/guarantee that project and module objects are singleton across their lifecycle.